// j-counter/lib/j-counter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JCounter = void 0;
var counter_1 = require("./counter");
var util_1 = require("util");
var summarizer_1 = require("./summarizer");
var _ = require("underscore");
var EXCEPTION_COUNTER_NOT_EXIST = 'The counter not exist';
var EXCEPTION_DELETE_DEFAULT_COUNTER = 'Can not delete the default counter';
// const enum Granularity {
//     DAY = 0,
//     HOUR,
//     MINUTE
// };
var JCounter = /** @class */ (function () {
    // granularity: Granularity;
    function JCounter() {
        this.counters = new Map();
        this.summarizers = new Map();
        this.DEFAULT_NAME = '__DEFAULT_COUNTER__';
        this.defaultCounter = new counter_1.Counter(this.DEFAULT_NAME);
        this.counters.set(this.DEFAULT_NAME, this.defaultCounter);
    }
    // setGranularity(granularity: number){
    //     this.granularity = granularity;
    // }
    JCounter.prototype.getCounter = function (name) {
        var cName = name || this.DEFAULT_NAME;
        var c = this.counters.get(cName);
        if (c === undefined)
            throw { message: EXCEPTION_COUNTER_NOT_EXIST };
        return c;
    };
    JCounter.prototype.createCounter = function (name) {
        if (this.counters.has(name)) {
            return this.counters.get(name);
        }
        var counter = new counter_1.Counter(name);
        this.counters.set(name, counter);
        return counter;
    };
    JCounter.prototype.getLastCount = function (name) {
        var c = this.getCounter(name);
        return c.lastCount;
    };
    JCounter.prototype.addOne = function (name) {
        var c = this.getCounter(name);
        c.increase();
        return true;
    };
    JCounter.prototype.add = function (pace, name) {
        if (!util_1.isNumber(pace))
            return false;
        var c = this.getCounter(name);
        return c.increase(pace);
    };
    JCounter.prototype.dropOne = function (name) {
        var c = this.getCounter(name);
        c.increase(-1);
    };
    JCounter.prototype.drop = function (pace, name) {
        if (!util_1.isNumber(pace))
            return false;
        var c = this.getCounter(name);
        return c.increase(-pace);
    };
    JCounter.prototype.getCountAt = function (ts, name) {
        var c = this.getCounter(name);
        return c.getCountAt(ts);
    };
    // //debug only
    // getKeys(name?: string){
    //     const c = this.getCounter(name);
    //     return c.keys;
    // }
    JCounter.prototype.getLastAt = function () {
        var c = this.getCounter(name);
        return c.lastAt;
    };
    JCounter.prototype.getPeak = function (begin, end, name) {
        var c = this.getCounter(name);
        return c.getPeak(begin, end);
    };
    JCounter.prototype.getAvg = function (begin, end, name) {
        var c = this.getCounter(name);
        return c.getAvg(begin, end);
    };
    JCounter.prototype.cut = function (cutAt, name) {
        var c = this.getCounter(name);
        return c.cut(cutAt);
    };
    JCounter.prototype.triggerSummarizer = function (interval, name) {
        var cName = name || this.DEFAULT_NAME;
        var c = this.getCounter(cName);
        var s = this.summarizers.get(cName);
        if (!s) {
            s = new summarizer_1.Summarizer(c);
            s.trigger(interval);
            this.summarizers.set(cName, s);
        }
        return s;
    };
    JCounter.prototype.getSummarizer = function (name) {
        var cName = name || this.DEFAULT_NAME;
        var s = this.summarizers.get(cName);
        if (s === undefined)
            throw { message: EXCEPTION_COUNTER_NOT_EXIST };
        return s;
    };
    // getSummary(name?: string){
    //     const cName = name || this.DEFAULT_NAME;
    //     const s = this.summarizers.get(cName);
    //     if (s === undefined) throw {message: EXCEPTION_COUNTER_NOT_EXIST};
    //     return s.getSummary();
    // }
    JCounter.prototype.getMinutePeak = function (name) {
        var cName = name || this.DEFAULT_NAME;
        var summarizer = this.summarizers.get(cName);
        if (summarizer === undefined)
            throw { message: EXCEPTION_COUNTER_NOT_EXIST };
        var summary = summarizer.getSummary();
        var vals = summary.values();
        var element = vals.next();
        var result = [];
        while (!element.done) {
            var hourSummary = element.value;
            result.push({ hour: hourSummary.hour, peaks: hourSummary.peaks });
            element = vals.next();
        }
        return result;
    };
    JCounter.prototype.getMinuteAvg = function (name) {
        var cName = name || this.DEFAULT_NAME;
        var summarizer = this.summarizers.get(cName);
        if (summarizer === undefined)
            throw { message: EXCEPTION_COUNTER_NOT_EXIST };
        var summary = summarizer.getSummary();
        var vals = summary.values();
        var element = vals.next();
        var result = [];
        while (!element.done) {
            var hourSummary = element.value;
            result.push({ hour: hourSummary.hour, avgs: hourSummary.avgs });
            element = vals.next();
        }
        return result;
    };
    JCounter.prototype.getHourPeak = function (name) {
        var cName = name || this.DEFAULT_NAME;
        var summarizer = this.summarizers.get(cName);
        if (summarizer === undefined)
            throw { message: EXCEPTION_COUNTER_NOT_EXIST };
        var summary = summarizer.getSummary();
        var vals = summary.values();
        var element = vals.next();
        var result = [];
        while (!element.done) {
            var hourSummary = element.value;
            result.push({ hour: hourSummary.hour, peak: _.max(hourSummary.peaks) });
            element = vals.next();
        }
        return result;
    };
    JCounter.prototype.getHourAvg = function (name) {
        var cName = name || this.DEFAULT_NAME;
        var summarizer = this.summarizers.get(cName);
        if (summarizer === undefined)
            throw { message: EXCEPTION_COUNTER_NOT_EXIST };
        var summary = summarizer.getSummary();
        var vals = summary.values();
        var element = vals.next();
        var result = [];
        while (!element.done) {
            var hourSummary = element.value;
            var sum = hourSummary.avgs.reduce(summarizer.sumReducer);
            result.push({ hour: hourSummary.hour, avg: sum / hourSummary.getLength() });
            element = vals.next();
        }
        return result;
    };
    JCounter.prototype.deleteCounter = function (name) {
        if (name !== this.DEFAULT_NAME) {
            var c = this.counters.get(name);
            if (c) {
                c.destroy();
                this.counters.delete(name);
            }
        }
        else {
            throw { message: EXCEPTION_DELETE_DEFAULT_COUNTER };
        }
        var s = this.summarizers.get(name);
        if (s) {
            s.destroy();
            this.summarizers.delete(name);
        }
    };
    JCounter.prototype.destroyAll = function () {
        this.counters.forEach(function (c) {
            c.destroy();
        });
        this.summarizers.forEach(function (s) {
            s.destroy();
        });
    };
    JCounter.prototype.pause = function (name) {
        var c = this.getCounter(name);
        c.recording = false;
    };
    JCounter.prototype.resume = function (name) {
        var c = this.getCounter(name);
        c.recording = true;
    };
    return JCounter;
}());
exports.JCounter = JCounter;
