// bz-db/dist/constants/config.js
const CONFIG = {
    //Block chain
    DIFFICULTY: 0,
    ROOT_BLOCK_HASH: 'b076b4ac5dfd570677538e23b54818022a379d2e8da1ef6f1b40f08965b528ff',
    HASH_ALGORITHM: 'sha256',
    HASH_KEY: 'a;ldskjf12o398230qhjasdfj;lav',
    CLEAN_CHAIN_KEEP_LENGTH: 100,
    CLEAN_BLOCK_KEEP_TIME: 1000 * 60 * 60 * 24 * 7,
    // CLEAN_BLOCK_KEEP_TIME: 1000 * 60 , // test only
    CLEAN_CHAIN_KEEP_LENGTH_MAX: 1000,
    MIN_TXN_REDUNDENT: 0,
    BLOCK_COMPETE_FACTOR_LENGTH: 10000,
    // Task
    NO_TIMEOUT: -1,
    TASK_DESC: 'Task id: ',
    TASK_TRACK_PERF: false,
    TASK_MAX_COUNT: 1000,
    // txn
    TXN_MAX_TRY: 3,
    TXN_WAIT_TOUCH: 1500,
    // Interval to auto checkin
    AUTO_CHECKIN_INTERVAL: 1000 * 60 * 60 * 2,
    // AUTO_CHECKIN_INTERVAL: 1000 * 60 * 2, // 2 minutes, only for testing.
    PULL_TOLERANCE_BLOCKS: 10,
    PULL_TOLERANCE_PERCENT: 0.2,
    ILLEGAL_CONNECTION_TOLERANCE: 300000,
    DATA_CONFLICT_PRINT_COUNT: 100,
    MESSAGE_COMPRESS: 'GZIP',
    PACKET_END_ESC: '[END]',
    UID_NUMBER_STR_RADIX: 36,
    UID_MAX_LEVEL: 4,
    UID_RANDOM_RANGE: 100,
    UID_BASE_TIME: Date.parse('2021-07-31'),
    UID_SEPRATOR: '-',
    UID_SEQUENCE_MAX: 10000,
    // log
    LOG_ENABLE_TIME_TRACE: false,
    // node
    NODE_DEFAULT_PORT: 8643,
    // CMD
    CMD_SHUTDOWN_TIMEOUT: 5000,
    //
    EXTEND_ENTITY: "extend_dataEntity",
    EXTEND_ENTITY_META: {
        name: 'extend_dataEntity',
        primaryKeys: ['name'],
        persistMethod: 1,
        filePath: '_extend'
    },
    // Performance
    EVENT_LOOP_BUSY_PERCENT: 0.65,
    EVENT_LOOP_CRAZY_BUSY_PERCENT: 0.9,
    EVENT_LOOP_SAMPLE_INTERVAL: 5000,
    MAXIMUM_DATA_CONFLICT_AMOUNT: 100,
    CHECK_NODE_STATE_INTERVAL_TIME: 60000,
    CHECK_NODE_STATE_DEBOUNCE_TIME: 5000 // 5 sec
};
export { CONFIG };
