// perf-tracker/src/tracker/event-loop-tracker.ts
import { EventLoopUtilization } from 'perf_hooks';
import { PerfTracker, EventLoopSampler } from '../index';
import { EventEmitter } from 'events';

export interface IEluConfig {
    busy: number
    crazy: number
    doTrackPercent: boolean
    doEmitAll: boolean
    interval: number
}

export const EluEvents = {
    BUSY: 'busy',
    NOTBUSY: 'notBusy',
    CRAZY: 'crazy',
    NOTCRAZY: 'notCrazy',
    BELOW_10: '<10',
    ABOVE_10: '>=10',
    BELOW_30: '<30',
    ABOVE_30: '>=30',
    BELOW_50: '<50',
    ABOVE_50: '>=50',
    BELOW_70: '<70',
    ABOVE_70: '>=70',
    BELOW_90: '<90',
    ABOVE_90: '>=90',
    SAMPLE: 'sample'
}

const ELU_NAME = 'elu'

/**
 * Event loop utilization tracker
 */
export class EventLoopTracker extends EventEmitter {

    private config: IEluConfig
    public latestElu: EventLoopUtilization;
    public isAbove10: boolean = false;
    public isAbove30: boolean = false;
    public isAbove50: boolean = false;
    public isAbove70: boolean = false;
    // public isAbove80: boolean = false;
    public isAbove90: boolean = false;
    public isBusy: boolean = false;
    public isCrazy: boolean = false;

    private eluTracker: PerfTracker;

    constructor(config: IEluConfig = {
        busy: 0.3,
        crazy: 0.9,
        doTrackPercent: true,
        doEmitAll: false,
        interval: 30000
    }) {
        super();
        this.config = config
        const eluTracker = new PerfTracker();
        eluTracker.addSampler(ELU_NAME, new EventLoopSampler(), this.config.interval);
        eluTracker.addBenchmark(ELU_NAME, (data) => {
            if (data === undefined) {
                return false;
            }
            const elu = data as EventLoopUtilization;
            if (elu.utilization !== undefined) {
                const eu: number = elu.utilization;
                this.setBusy(eu);
                this.setPercent(eu);
                this.emitSample(elu);
            }
            return false;
        }, () => {
            // The logics are covered in benchmarkFunc, so doing nothing here.
        });
        eluTracker.startSampler(ELU_NAME);
        this.eluTracker = eluTracker;
    }

    stop() {
        this.eluTracker.stopSampler(ELU_NAME);
    }

    emitSample(elu: EventLoopUtilization) {
        if (this.config.doEmitAll) {
            this.emit(EluEvents.SAMPLE, elu);
        }
    }

    setBusy(eu: number) {
        if (eu >= this.config.busy) {
            if (!this.isBusy) {
                this.isBusy = true;
                this.emit(EluEvents.BUSY, eu);
            }
        } else {
            if (this.isBusy) {
                this.isBusy = false;
                this.emit(EluEvents.NOTBUSY, eu);
            }
        }
        if (eu >= this.config.crazy) {
            if (!this.isCrazy) {
                this.isCrazy = true;
                this.emit(EluEvents.CRAZY, eu);
            }
        } else {
            if (this.isCrazy) {
                this.isCrazy = false;
                this.emit(EluEvents.NOTCRAZY, eu);
            }
        }
    }

    setPercent(eu: number) {
        if (!this.config.doTrackPercent) {
            return;
        }
        if (eu >= 0.1) {
            if (!this.isAbove10) {
                this.isAbove10 = true;
                this.emit(EluEvents.ABOVE_10, eu)
            }
        } else {
            if (this.isAbove10) {
                this.isAbove10 = false;
                this.emit(EluEvents.BELOW_10, eu)
            }
        }
        if (eu >= 0.3) {
            if (!this.isAbove30) {
                this.isAbove30 = true;
                this.emit(EluEvents.ABOVE_30, eu)
            }
        } else {
            if (this.isAbove30) {
                this.isAbove30 = false;
                this.emit(EluEvents.BELOW_30, eu)
            }
        }
        if (eu >= 0.5) {
            if (!this.isAbove50) {
                this.isAbove50 = true;
                this.emit(EluEvents.ABOVE_50, eu)
            }
        } else {
            if (this.isAbove50) {
                this.isAbove50 = false;
                this.emit(EluEvents.BELOW_50, eu)
            }
        }
        if (eu >= 0.7) {
            if (!this.isAbove70) {
                this.isAbove70 = true;
                this.emit(EluEvents.ABOVE_70, eu)
            }
        } else {
            if (this.isAbove70) {
                this.isAbove70 = false;
                this.emit(EluEvents.BELOW_70, eu)
            }
        }
        if (eu >= 0.9) {
            if (!this.isAbove90) {
                this.isAbove90 = true;
                this.emit(EluEvents.ABOVE_90, eu)
            }
        } else {
            if (this.isAbove90) {
                this.isAbove90 = false;
                this.emit(EluEvents.BELOW_90, eu)
            }
        }
    }
}
