// bz-db/src/constants/channels.ts
const CHANNELS = {
    CHAT: '/chat',
    STATUS: '/status',
    STATUS_CONFIRM: '/status-confirm',
    CHECKIN: '/checkin',
    CHECKIN_RES: '/checkin-res',
    TXN: '/txn',
    TXN_CONFIRM: '/txn-confirm',
    TXN_GO: '/txn-go',
    TXN_DONE: '/txn-done',
    BATCH_TXN: '/batch-txn',
    BATCH_TXN_CONFIRM: '/batch-txn-confirm',
    BATCH_TXN_GO: '/batch-txn-go',
    BATCH_TXN_DONE: '/batch-txn-done',
    SELECT_MEMORY: '/select-memory',
    SELECT_MEMORY_DONE: '/select-memory-done',
    PULL: '/pull',
    PULL_RES: '/pull-res',
    PUSH_GO: '/push-go',
    PUSH_DONE: '/push-done',
    FORCEPULL_GO: '/force-pull-go',
    FORCEPULL_DONE: '/force-pull-down',
    REQ_CHECKIN: '/req-checkin',
    REQ_CHECKIN_ACK: '/req-checkin-ack',
    INTRODUCE: '/introduce',
    KICK: '/kick',
    HAPPY_TO_JOIN:'/happy-to-join',
    LEAVE: '/leave',
    PING:'/ping',
    PONG:'/pong',
    FILE_TRANSFER_REQ: '/file-transfer-req',
    FILE_TRANSFER_RES: '/file-transfer-res',
    FILE_TRANSFER_PULL_DONE: '/file-transfer-done',
    CMD: '/cmd',
    CMD_RES: '/cmd-res',
    REJECT: '/reject',
    AUTO_SCALE: '/auto-scale',
    AUTO_SCALE_RES: '/auto-scale-res',
    AUTO_RESOLVE: '/auto-resolve-go',
    AUTO_RESOLVE_DONE: '/auto-resolve-done'
}

export {
    CHANNELS
}
