// bz-db/src/constants/config.ts

const CONFIG = {
    //Block chain
    DIFFICULTY: 0,
    ROOT_BLOCK_HASH: 'b076b4ac5dfd570677538e23b54818022a379d2e8da1ef6f1b40f08965b528ff',

    HASH_ALGORITHM: 'sha256',
    HASH_KEY: 'a;ldskjf12o398230qhjasdfj;lav',

    CLEAN_CHAIN_KEEP_LENGTH: 100, // Keep at least 100 blocks when clear chain
    CLEAN_BLOCK_KEEP_TIME: 1000 * 60 * 60 * 24 * 7, // Keep block for at least 1 week
    // CLEAN_BLOCK_KEEP_TIME: 1000 * 60 , // test only
    CLEAN_CHAIN_KEEP_LENGTH_MAX: 1000, // Keep 1000 blocks at most, even the block is created within a week

    MIN_TXN_REDUNDENT: 0,

    BLOCK_COMPETE_FACTOR_LENGTH: 10000,

    // Task
    NO_TIMEOUT: -1,
    TASK_DESC: 'Task id: ',
    TASK_TRACK_PERF: false, // Test purpose only, this will output the life time of each task.
    TASK_MAX_COUNT: 1000,

    // txn
    TXN_MAX_TRY: 3,
    TXN_WAIT_TOUCH: 1500, // Wait for a period before update the block index.

    // Interval to auto checkin
    AUTO_CHECKIN_INTERVAL: 1000 * 60 * 60 * 2, // 2 hours
    // AUTO_CHECKIN_INTERVAL: 1000 * 60 * 2, // 2 minutes, only for testing.

    PULL_TOLERANCE_BLOCKS: 10, // Do full pull if the local chain missed more than this number of blocks, other wize append the missed blocks
    PULL_TOLERANCE_PERCENT: 0.2, // Do full pull if the local chain missed more than this number of blocks, other wize append the missed blocks

    ILLEGAL_CONNECTION_TOLERANCE: 300000, // Kill the connection from unknown node after 5 minutes
    
    DATA_CONFLICT_PRINT_COUNT: 100,
    MESSAGE_COMPRESS: 'GZIP', // 'GZIP' or any other value
    PACKET_END_ESC: '[END]',

    UID_NUMBER_STR_RADIX: 36, // Radix of timestamp output, 2 - 36
    UID_MAX_LEVEL: 4, // Max level of uid
    UID_RANDOM_RANGE: 100, // Range for randome number generation
    UID_BASE_TIME: Date.parse('2021-07-31'), // The base ts
    UID_SEPRATOR: '-', // Seprator of the parts of unique ID
    UID_SEQUENCE_MAX: 10000, // Rotates the internal sequence once it exceeds this number

    // log
    LOG_ENABLE_TIME_TRACE: false,

    // node
    NODE_DEFAULT_PORT: 8643,

    // CMD
    CMD_SHUTDOWN_TIMEOUT: 5000,

    //
    EXTEND_ENTITY: "extend_dataEntity",

    EXTEND_ENTITY_META:{
        name: 'extend_dataEntity',
        primaryKeys: ['name'],
        persistMethod: 1, //PersistMethod.PERSIST_METHOD_LIST_FILE,
        filePath: '_extend'
    },

    // Performance
    EVENT_LOOP_BUSY_PERCENT: 0.65, // checkin() will be postponed upon this benchmark
    EVENT_LOOP_CRAZY_BUSY_PERCENT: 0.9, // introduceNode, transactions and forcepull will be rejected upon this benchmark
    EVENT_LOOP_SAMPLE_INTERVAL: 5000, // event loop utilization sampling interval. Make sure this is not too frequent and impact the performance

    MAXIMUM_DATA_CONFLICT_AMOUNT: 100,
    
    CHECK_NODE_STATE_INTERVAL_TIME: 60000,  // 1 min
    
    CHECK_NODE_STATE_DEBOUNCE_TIME: 5000    // 5 sec
    
}

export {
    CONFIG
}
